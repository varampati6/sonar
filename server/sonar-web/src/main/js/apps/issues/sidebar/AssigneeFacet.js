/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import { sortBy, uniq, without } from 'lodash';
import FacetBox from './components/FacetBox';
import FacetHeader from './components/FacetHeader';
import FacetItem from './components/FacetItem';
import FacetItemsList from './components/FacetItemsList';
import FacetFooter from './components/FacetFooter';
import { searchAssignees } from '../utils';
import type { ReferencedUser, Component } from '../utils';
import Avatar from '../../../components/ui/Avatar';
import { translate } from '../../../helpers/l10n';

type Props = {|
  assigned: boolean,
  assignees: Array<string>,
  component?: Component,
  facetMode: string,
  onChange: (changes: {}) => void,
  onToggle: (property: string) => void,
  open: boolean,
  stats?: { [string]: number },
  referencedUsers: { [string]: ReferencedUser }
|};

export default class AssigneeFacet extends React.PureComponent {
  props: Props;

  static defaultProps = {
    open: true
  };

  property = 'assignees';

  handleItemClick = (itemValue: string) => {
    if (itemValue === '') {
      // unassigned
      this.props.onChange({ assigned: !this.props.assigned, assignees: [] });
    } else {
      // defined assignee
      const { assignees } = this.props;
      const newValue = sortBy(
        assignees.includes(itemValue) ? without(assignees, itemValue) : [...assignees, itemValue]
      );
      this.props.onChange({ assigned: true, assignees: newValue });
    }
  };

  handleHeaderClick = () => {
    this.props.onToggle(this.property);
  };

  handleClear = () => {
    this.props.onChange({ assigned: true, assignees: [] });
  };

  handleSearch = (query: string) => searchAssignees(query, this.props.component);

  handleSelect = (assignee: string) => {
    const { assignees } = this.props;
    this.props.onChange({ assigned: true, [this.property]: uniq([...assignees, assignee]) });
  };

  isAssigneeActive(assignee: string) {
    return assignee === '' ? !this.props.assigned : this.props.assignees.includes(assignee);
  }

  getAssigneeName(assignee: string): React.Element<*> | string {
    if (assignee === '') {
      return translate('unassigned');
    } else {
      const { referencedUsers } = this.props;
      if (referencedUsers[assignee]) {
        return (
          <span>
            <Avatar
              className="little-spacer-right"
              hash={referencedUsers[assignee].avatar}
              name={referencedUsers[assignee].name}
              size={16}
            />
            {referencedUsers[assignee].name}
          </span>
        );
      } else {
        return assignee;
      }
    }
  }

  getStat(assignee: string): ?number {
    const { stats } = this.props;
    return stats ? stats[assignee] : null;
  }

  renderOption = (option: { avatar: string, label: string }) => {
    return (
      <span>
        {option.avatar != null &&
          <Avatar
            className="little-spacer-right"
            hash={option.avatar}
            name={option.label}
            size={16}
          />}
        {option.label}
      </span>
    );
  };

  renderList() {
    const { stats } = this.props;

    if (!stats) {
      return null;
    }

    const assignees = sortBy(
      Object.keys(stats),
      // put unassigned first
      key => (key === '' ? 0 : 1),
      // the sort by number
      key => -stats[key]
    );

    return (
      <FacetItemsList>
        {assignees.map(assignee => (
          <FacetItem
            active={this.isAssigneeActive(assignee)}
            facetMode={this.props.facetMode}
            key={assignee}
            name={this.getAssigneeName(assignee)}
            onClick={this.handleItemClick}
            stat={this.getStat(assignee)}
            value={assignee}
          />
        ))}
      </FacetItemsList>
    );
  }

  renderFooter() {
    if (!this.props.stats) {
      return null;
    }

    return (
      <FacetFooter
        onSearch={this.handleSearch}
        onSelect={this.handleSelect}
        renderOption={this.renderOption}
      />
    );
  }

  render() {
    return (
      <FacetBox property={this.property}>
        <FacetHeader
          name={translate('issues.facet', this.property)}
          onClear={this.handleClear}
          onClick={this.handleHeaderClick}
          open={this.props.open}
          values={this.props.assignees.length + (this.props.assigned ? 0 : 1)}
        />

        {this.props.open && this.renderList()}
        {this.props.open && this.renderFooter()}
      </FacetBox>
    );
  }
}
