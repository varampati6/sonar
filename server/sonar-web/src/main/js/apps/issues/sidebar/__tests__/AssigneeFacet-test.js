/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import AssigneeFacet from '../AssigneeFacet';

jest.mock('../../../../store/rootReducer', () => ({}));

const renderAssigneeFacet = (props?: {}) =>
  shallow(
    <AssigneeFacet
      assigned={true}
      assignees={[]}
      facetMode="count"
      onChange={jest.fn()}
      onToggle={jest.fn()}
      open={true}
      referencedUsers={{ foo: { avatar: 'avatart-foo', name: 'name-foo' } }}
      stats={{ '': 5, foo: 13, bar: 7 }}
      {...props}
    />
  );

it('should render', () => {
  expect(renderAssigneeFacet()).toMatchSnapshot();
});

it('should render without stats', () => {
  expect(renderAssigneeFacet({ stats: null })).toMatchSnapshot();
});

it('should select unassigned', () => {
  expect(renderAssigneeFacet({ assigned: false })).toMatchSnapshot();
});

it('should select user', () => {
  expect(renderAssigneeFacet({ assignees: ['foo'] })).toMatchSnapshot();
});

it('should render footer select option', () => {
  const wrapper = renderAssigneeFacet();
  expect(
    wrapper.instance().renderOption({ avatar: 'avatar-foo', label: 'name-foo' })
  ).toMatchSnapshot();
});

it('should call onChange', () => {
  const onChange = jest.fn();
  const wrapper = renderAssigneeFacet({ assignees: ['foo'], onChange });
  const itemOnClick = wrapper.find('FacetItem').first().prop('onClick');

  itemOnClick('');
  expect(onChange).lastCalledWith({ assigned: false, assignees: [] });

  itemOnClick('bar');
  expect(onChange).lastCalledWith({ assigned: true, assignees: ['bar', 'foo'] });

  itemOnClick('foo');
  expect(onChange).lastCalledWith({ assigned: true, assignees: [] });
});

it('should call onToggle', () => {
  const onToggle = jest.fn();
  const wrapper = renderAssigneeFacet({ onToggle });
  const headerOnClick = wrapper.find('FacetHeader').prop('onClick');

  headerOnClick();
  expect(onToggle).lastCalledWith('assignees');
});

it('should handle footer callbacks', () => {
  const onChange = jest.fn();
  const wrapper = renderAssigneeFacet({ assignees: ['foo'], onChange });
  const onSelect = wrapper.find('FacetFooter').prop('onSelect');

  onSelect('qux');
  expect(onChange).lastCalledWith({ assigned: true, assignees: ['foo', 'qux'] });
});
