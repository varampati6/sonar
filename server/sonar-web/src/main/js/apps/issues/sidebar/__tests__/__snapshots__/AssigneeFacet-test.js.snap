// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<FacetBox
  property="assignees"
>
  <FacetHeader
    name="issues.facet.assignees"
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={0}
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={false}
      facetMode="count"
      halfWidth={false}
      name="unassigned"
      onClick={[Function]}
      stat={5}
      value=""
    />
    <FacetItem
      active={false}
      disabled={false}
      facetMode="count"
      halfWidth={false}
      name={
        <span>
          <Connect(Avatar)
            className="little-spacer-right"
            hash="avatart-foo"
            name="name-foo"
            size={16}
          />
          name-foo
        </span>
      }
      onClick={[Function]}
      stat={13}
      value="foo"
    />
    <FacetItem
      active={false}
      disabled={false}
      facetMode="count"
      halfWidth={false}
      name="bar"
      onClick={[Function]}
      stat={7}
      value="bar"
    />
  </FacetItemsList>
  <FacetFooter
    onSearch={[Function]}
    onSelect={[Function]}
    renderOption={[Function]}
  />
</FacetBox>
`;

exports[`should render footer select option 1`] = `
<span>
  <Connect(Avatar)
    className="little-spacer-right"
    hash="avatar-foo"
    name="name-foo"
    size={16}
  />
  name-foo
</span>
`;

exports[`should render without stats 1`] = `
<FacetBox
  property="assignees"
>
  <FacetHeader
    name="issues.facet.assignees"
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={0}
  />
</FacetBox>
`;

exports[`should select unassigned 1`] = `
<FacetBox
  property="assignees"
>
  <FacetHeader
    name="issues.facet.assignees"
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={1}
  />
  <FacetItemsList>
    <FacetItem
      active={true}
      disabled={false}
      facetMode="count"
      halfWidth={false}
      name="unassigned"
      onClick={[Function]}
      stat={5}
      value=""
    />
    <FacetItem
      active={false}
      disabled={false}
      facetMode="count"
      halfWidth={false}
      name={
        <span>
          <Connect(Avatar)
            className="little-spacer-right"
            hash="avatart-foo"
            name="name-foo"
            size={16}
          />
          name-foo
        </span>
      }
      onClick={[Function]}
      stat={13}
      value="foo"
    />
    <FacetItem
      active={false}
      disabled={false}
      facetMode="count"
      halfWidth={false}
      name="bar"
      onClick={[Function]}
      stat={7}
      value="bar"
    />
  </FacetItemsList>
  <FacetFooter
    onSearch={[Function]}
    onSelect={[Function]}
    renderOption={[Function]}
  />
</FacetBox>
`;

exports[`should select user 1`] = `
<FacetBox
  property="assignees"
>
  <FacetHeader
    name="issues.facet.assignees"
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={1}
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={false}
      facetMode="count"
      halfWidth={false}
      name="unassigned"
      onClick={[Function]}
      stat={5}
      value=""
    />
    <FacetItem
      active={true}
      disabled={false}
      facetMode="count"
      halfWidth={false}
      name={
        <span>
          <Connect(Avatar)
            className="little-spacer-right"
            hash="avatart-foo"
            name="name-foo"
            size={16}
          />
          name-foo
        </span>
      }
      onClick={[Function]}
      stat={13}
      value="foo"
    />
    <FacetItem
      active={false}
      disabled={false}
      facetMode="count"
      halfWidth={false}
      name="bar"
      onClick={[Function]}
      stat={7}
      value="bar"
    />
  </FacetItemsList>
  <FacetFooter
    onSearch={[Function]}
    onSelect={[Function]}
    renderOption={[Function]}
  />
</FacetBox>
`;
